aten/src/ATen/native/cudnn/ConvShared.cpp

```diff
@@ -0,0 +1,500 @@
+// code shared by v7 and v8. Conv_v7.cpp contains implementation of
+// convolution using cuDNN v7 API. Conv_v8.cpp contains implementation
+// with v8 API.
+//
+// NOTE [ Convolution design ]
+//
+// cuDNN convolutions does not handle bias. Bias is handled outside.
+//
+// The general strategy:
+//
+//    - cudnn_convolution (Tensor)
+//      Entry points for clients
+//
+//    - cudnn_convolution_forward (TensorArg)
+//      Entry point, which may be reused between regular
+//      convolution and transposed convolution.
+//
+//    - raw_cudnn_convolution_forward_out (Tensor)
+//      Function that has different implementation on Conv_v7.cpp
+//      and Conv_v8.cpp
+//
+// The raw API directly invokes CuDNN and are implemeted differently
+// on cuDNN v7 and cuDNN v8
+//
+// There are a few reasons this should never be directly exposed
+// via ATen:
+//
+//    - It takes output as a parameter (this should be computed!)
+//    - It doesn't do input checking
+//    - It doesn't resize output (it is assumed to be correctly sized)
+//
+// Where does argument checking happen?  Here's the division of
+// responsibility:
+//  - Things that happen in at::Tensor
+//    - TensorArg allocation
+//  - Things that happen in TensorArg
+//    - Check arguments (type, GPU, shape)
+
+namespace at { namespace native {
+
+// ---------------------------------------------------------------------
+//
+// ConvolutionParams and ConvolutionArgs
+//
+// ---------------------------------------------------------------------
+
+std::ostream& operator<<(std::ostream & out, const ConvolutionParams& params) {
+  out << "ConvolutionParams \n"
+    << "    data_type = " << cudnnTypeToString(params.dataType) << "\n"
+    << "    padding = " << ArrayRef<int>{params.padding} << "\n"
+    << "    stride = " << ArrayRef<int>{params.stride} << "\n"
+    << "    dilation = " << ArrayRef<int>{params.dilation} << "\n"
+    << "    groups = " << params.groups << "\n"
+    << "    deterministic = " << (params.deterministic ? "true" : "false") << "\n"
+    << "    allow_tf32 = " << (params.allow_tf32 ? "true" : "false") << "\n";
+
+  return out;
+}
+
+// NB: This can't be a constructor, because then ConvolutionParams
+// would not be a POD anymore.
+// TODO: Use TensorGeometry here instead of the entire Tensor, which we
+// don't actually need.  (OTOH: We can always pass in
+// grad_input/grad_output, so this is not very pressing)
+void setConvolutionParams(
+    ConvolutionParams* params,
+    const at::Tensor& input, const at::Tensor& weight,
+    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation,
+    int64_t groups, bool deterministic, bool allow_tf32) {
+
+  cudnnDataType_t dataType = getCudnnDataType(input);
+  memset(params, 0, sizeof(ConvolutionParams));
+  params->dataType = dataType;
+  // ASSERT(weight.dim() == input.dim())
+  for (int i = 0; i != input.dim(); ++i) {
+    params->input_size[i] = (int) input.size(i);
```

Comment:
Can you please do unchecked accesses here? `input.sizes()[i]`

Reply:
will fix in later PR

Reply:
IIRC, these were all from the old code, right?

Reply:
Yes, they are, but that's not a reason not to fix them.